### PR TITLE
support trainable Sum observables

### LINKED ISSUE #4199: [BUG] Issue with the gradients of the expectation value of an operator
### Expected behavior

Hi, I'm trying to compute the gradient of the circuit with respect to the parameters on the operator $\theta_{i,i+1}(\sigma^+_i\sigma^-_{i+1} + \sigma^+_{i+1}\sigma^-_i)$ where $\theta$ is my variable. When I compute the expectation value of the circuit, I can see that individual terms of my operator are differentiable, e.g. for $\theta_0(\sigma^+_0\sigma^-_1 + \sigma^+_1\sigma^-_0)$ I can get a gradient with respect to $\theta_0$ but as soon as I add other terms I get None which I was expecting to get the gradient for every term.

### Actual behavior

Gradient returns ``None``.

### Additional information

I get zero for the gradient when I use autograd or Jax.

### Source code

```shell
import pennylane as qml
import tensorflow as tf
import pennylane.numpy as np

# here s_prod or sum is not necessary +, @ and * works the same
sig_plus = lambda w: qml.s_prod(0.5, (qml.sum(qml.PauliX(w) , qml.s_prod(1j, qml.PauliY(w)))))
sig_minus = lambda w: qml.s_prod(0.5, (qml.sum(qml.PauliX(w) , qml.s_prod(-1j, qml.PauliY(w)))))
term = lambda w: qml.sum(qml.prod(sig_plus(w) , sig_minus(w + 1)) , qml.prod(sig_plus(w + 1), sig_minus(w)))

@qml.qnode(qml.device("default.qubit.tf", wires=4), interface="tf")
def circ(phi, theta):
    # This portion of the circuit is random not important
    qml.RY(phi[0], 0)
    qml.RY(phi[1], 1)
    qml.RY(phi[2], 2)
    qml.CNOT([0,1])
    qml.CNOT([1,2])
    ###########################
    return qml.expval(theta[0]*term(0) + theta[1]*term(1) ) # only 0 or only 1 works just fine, also tried with qml.sum and qml.s_prod

theta = tf.Variable(np.random.uniform(0,1,3))
phi = tf.Variable(np.random.uniform(0,1,3))

def cost(phi, theta):
    return circ(phi, theta)

with tf.GradientTape() as tape:
    loss = cost(phi, theta)
tape.gradient(loss, [phi, theta])

#[<tf.Tensor: shape=(3,), dtype=float64, numpy=array([0.00131378, 0.00458278, 0.00750094])>,
# None]
```

### Tracebacks

```shell
No error has been generated
```

### System information

```shell
TF version 2.11.0
TF compiler version Apple LLVM 13.1.6 (clang-1316.0.21.1)

Name: PennyLane
Version: 0.30.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/jackaraz/packages/miniconda3/lib/python3.9/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml
Required-by: field-theory, PennyLane-Lightning, PennyLane-qiskit, qhbm

Platform info:           macOS-13.3.1-arm64-arm-64bit
Python version:          3.9.16
Numpy version:           1.23.5
Scipy version:           1.10.0
Installed devices:
- default.gaussian (PennyLane-0.30.0)
- default.mixed (PennyLane-0.30.0)
- default.qubit (PennyLane-0.30.0)
- default.qubit.autograd (PennyLane-0.30.0)
- default.qubit.jax (PennyLane-0.30.0)
- default.qubit.tf (PennyLane-0.30.0)
- default.qubit.torch (PennyLane-0.30.0)
- default.qutrit (PennyLane-0.30.0)
- null.qubit (PennyLane-0.30.0)
- lightning.qubit (PennyLane-Lightning-0.30.0)
- qiskit.aer (PennyLane-qiskit-0.28.0)
- qiskit.basicaer (PennyLane-qiskit-0.28.0)
- qiskit.ibmq (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.circuit_runner (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.sampler (PennyLane-qiskit-0.28.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Sum observables should be able to do anything that Hamiltonians can do. It turns out that computing the gradient of parameters on Sum observables did not work.

**Description of the Change:**
- Updates DefaultQubit to compute the expectation of Sums using the new `qml.devices.qubit.measure` to maintain trainability
- Updates `qml.devices.qubit.measure.get_measurement_function` to _always_ return `sum_of_terms_method` for Sums in backprop mode
 
A few little changes had to be made along the way for this to work with all autodiff frameworks:
- `obs.eigvals()` would be returned as a tracer if the parameters are trainable, but we called `np.array()` on it. Now I use `qml.math.asarray()` instead to avoid breaking trainability
- Torch tracers can't do scalar multiplication with numpy arrays, so that broke `SProd.eigvals` with a traced scalar. I use `qml.math.convert_like` to fix this, because it has custom code for exactly this

**Benefits:**
- We can compute the gradient of Sum observables!
- Very little code had to be added

**Possible Drawbacks:**
- only backprop support for now
- finite-shots does not work

**Related GitHub Issues:**
Fixes #4199 

**Future work**
- Support the things not yet supported (see the drawbacks section)
- Improve the choice made in get_measurement_function. Eg. if the Sum has no trainable parameters on it, this always-sum-of-terms solution might not be necessary

### CODE REVIEW

File: pennylane/devices/qubit/measure.py [hunk 1]
```diff
@@ -109,18 +109,24 @@ def get_measurement_function(
...
 
-            if isinstance(measurementprocess.obs, Hamiltonian) or (
-                isinstance(measurementprocess.obs, Sum)
-                and measurementprocess.obs.has_overlapping_wires
-                and len(measurementprocess.obs.wires) > 7
-            ):
-                # Use tensor contraction for `Sum` expectation values with non-commuting summands
-                # and 8 or more wires as it's faster than using eigenvalues.
-
+            backprop_mode = math.get_interface(state) != "numpy"
```
Comment (albi3ro) [inline pennylane/devices/qubit/measure.py:112]:
```suggestion
            backprop_mode = math.get_interface(state, *measurementprocess.obs.data) != "numpy"
```
Might also be something we need to consider. The state could still be numpy but the observable could have trainable parameters.

Comment (timmysilv) [reply pennylane/devices/qubit/measure.py:112]: I think it hasn't been a problem because we always [cast the state](https://github.com/PennyLaneAI/pennylane/blob/a60cf2074baae32d9b725c802886a16a0c7cdb62/pennylane/devices/default_qubit.py#L681) in the old DefaultQubit, but this will be good to have because we probably would have hit a bug in the future since we [aren't casting the state in new devices](https://github.com/PennyLaneAI/pennylane/blob/a60cf2074baae32d9b725c802886a16a0c7cdb62/pennylane/devices/qubit/simulate.py#L58) (yet, just need to pass the appropriate "like" value)

Comment (timmysilv) [reply pennylane/devices/qubit/measure.py:138]: i'll do this in a separate PR after to keep this one simple.

**Review by mudit2812 — APPROVED** [review]

**Comment by timmysilv** [comment]
[sc-39422]

File: tests/ops/op_math/test_sum.py [hunk 2]
```diff
@@ -1062,7 +1062,7 @@ def circuit(weights):
         assert qnp.allclose(grad, true_grad)
 
     def test_non_hermitian_op_in_measurement_process(self):
-        """Test that non-hermitian ops in a measurement process will raise a warning."""
+        """Test that non-hermitian ops in a measurement process will fail."""
```
Comment (albi3ro) [inline tests/ops/op_math/test_sum.py:1065]: If we specify `interface=None` on the qnode, then it will execute.  The only reason it was failing was because it was using the autograd device and assuming we needed to differentiate it.

**Review by albi3ro — APPROVED** [review]
Optionally reverting the test to such a form that it will execute.

File: tests/ops/op_math/test_sum.py (hunk 2 above)
Comment (timmysilv) [reply tests/ops/op_math/test_sum.py]: nice one. updated
